Allow to use self hosted Gitlab and add Project ID support.

b0c1 opened the pull request:
# Description

Separate GitLab URL for the project because some self-hosted GitLab not installed on the root, and added project id support (our GitLab did not work with project names)

## Issues Resolved

#337 

## Check List

- [x] Manual test on gitlab.com and self-hosted GitLab.
- [ ] New functionality has been documented in the README if applicable.
- [ ] Translations needed in other languages

johannesjo commented:
Excellent work! Thank you very much! 

One thing that might be missing: A little migration script to assign the default value inside `src/app/features/project/migrate-projects-state.util.ts`, because otherwise the base url might also be `undefined` instead of `null`. Alternatively we can change the code around line 217 to something like that:
``` 
    const base: string = projectConfig.gitlabBaseUrl || GITLAB_BASE_URL;
    const fixedUrl = base.match(/.*\/$/)
      ? projectConfig.gitlabBaseUrl
      : `${projectConfig.gitlabBaseUrl}/`;
    apiURL = fixedUrl + 'api/v4/projects/';
```


Maybe we can also add "Custom" to the form field caption to indicate that otherwise a default is used?

b0c1 commented:
> 
> 
> Excellent work! Thank you very much!
> 
> One thing that might be missing: A little migration script to assign the default value inside `src/app/features/project/migrate-projects-state.util.ts`, because otherwise the base url might also be `undefined` instead of `null`, which leads to an error. Alternatively we can change the code around line 217 to something like that:
> 
> ```
>     const base: string = projectConfig.gitlabBaseUrl || GITLAB_BASE_URL;
>     const fixedUrl = base.match(/.*\/$/)
>       ? projectConfig.gitlabBaseUrl
>       : `${projectConfig.gitlabBaseUrl}/`;
>     apiURL = fixedUrl + 'api/v4/projects/';
> ```
> 
> And maybe we can also add "Custom" to the form field caption to indicate that otherwise a default is used?

Ok, I changed the caption and replaced it with ```!==``` to ```!=``` because it will match to ```undefined``` too.

johannesjo commented:
Alright! Thank you very much!
